state/store/defaultstore/message.go

```diff
@@ -73,68 +73,129 @@ func (s *Message) MaxMessages() int {
 	return s.maxMsgs
 }
 
-func (s *Message) MessageSet(message discord.Message) error {
+func (s *Message) MessageSet(message discord.Message, update bool) error {
+	if s.maxMsgs <= 0 {
+		return nil
+	}
+
 	iv, _ := s.channels.LoadOrStore(message.ChannelID)
 
 	msgs := iv.(*messages)
 
 	msgs.mut.Lock()
 	defer msgs.mut.Unlock()
 
-	for i, m := range msgs.messages {
-		if m.ID == message.ID {
-			DiffMessage(message, &m)
-			msgs.messages[i] = m
-			return nil
+	if update {
+		// Opt for a linear latest-to-oldest search in favor of something like
+		// sort.Search, since more recent messages are more likely to be edited
+		// than older ones.
+		for i, oldMessage := range msgs.messages {
+			// We found a match, update it.
+			if oldMessage.ID == message.ID {
+				DiffMessage(message, &oldMessage)
+				msgs.messages[i] = oldMessage // Now updated.
+				return nil
+			}
 		}
+
+		return nil
 	}
 
-	// Order: latest to earliest, similar to the API.
+	switch {
+	case len(msgs.messages) == 0:
+		msgs.messages = []discord.Message{message}
+	case shouldPrependMessage(message, msgs.messages):
+		if len(msgs.messages) == s.maxMsgs {
+			copy(msgs.messages[1:], msgs.messages)
+			msgs.messages[0] = message
+		} else {
+			msgs.messages = append([]discord.Message{message}, msgs.messages...)
+		}
+	case shouldAppendMessage(message, msgs.messages) && len(msgs.messages) < s.maxMsgs:
+		msgs.messages = append(msgs.messages, message)
+	}
 
-	// Check if we already have the message. Try to derive the order otherwise.
-	var insertAt int
+	// We already have this message or we can't append any more messages.
+	return nil
+}
 
-	// Since we make the order guarantee ourselves, we can trust that we're
-	// iterating from latest to earliest.
-	for insertAt < len(msgs.messages) {
-		// Check if the new message is older. If it is, then we should insert it
-		// right after this message (or before this message in the list; i-1).
-		if message.ID > msgs.messages[insertAt].ID {
-			break
+// shouldPrependMessage checks if the passed message may be prepended to the
+// passed message slice, that is ordered from latest to oldest.
+//
+// shouldPrependMessage is biased as it will return true if the timestamps
+// of the passed message and the latest message match, even though the true
+// order cannot be determined in that case.
+func shouldPrependMessage(message discord.Message, messages []discord.Message) bool {
```

Comment:
Wouldn't it be more ideal to have an API that returns the index to be inserted in the list? `-1` can be returned to mean ignore.

Reply:
Insertion into the slice, as in somewhere in the middle, is never desired though, since the messages stored are guaranteed to be complete. The only reason there is even a check for duplicates, is to prevent toctou problems where in-between the time of checking and calling `MessageSet` there could be another function that already added that message. Gaps, however, are guaranteed not to occur, which is why I don't think we even need indexes.

Reply:
That's a fair point, though I still think we could return a number to indicate prepending or appending without needing two functions. Something like `1` for append, `-1` for prepend, and `0` for ignore?

Reply:
Sure, I'll get on doing that tomorrow.